kernel/signal.c

```diff
@@ -238,27 +271,48 @@ static long _handle_one_signal(
             thread->signal.mask |= mask;
 
         // ATTN: handle other signal flags, e.g., SA_NOCLDSTOP, SA_NOCLDWAIT,
-        // SA_ONSTACK, SA_RESETHAND, SA_RESTART, etc.
+        // SA_RESETHAND, SA_RESTART, etc.
 
         /* save the original fsbase */
         void* original_fsbase = myst_get_fsbase();
+        stack_t* altstack = &thread->signal.altstack;
 
         /* restore the user-space fsbase, which is pthread_self() */
         myst_set_fsbase(thread->crt_td);
 
+        use_alt_stack = (action->flags & SA_ONSTACK) &&
+                        !(altstack->ss_flags & (SS_DISABLE | SS_ONSTACK)) &&
+                        (altstack->ss_sp != 0);
+
         if ((action->flags & SA_SIGINFO) != 0)
         {
-            ((sigaction_function_t)(action->handler))(
-                signum, siginfo, &context);
-            // Copy back mcontext (register states)
-            if (mcontext != NULL)
-                *mcontext = context.uc_mcontext;
+            arg.sigaction_handler = (sigaction_function_t)(action->handler);
+            arg.signum = signum;
+            arg.siginfo = siginfo;
+            arg.ucontext = &context;
         }
         else
         {
-            ((sigaction_handler_t)(action->handler))(signum);
+            arg.signum_handler = (sigaction_handler_t)(action->handler);
+            arg.signum = signum;
         }
 
+        if (use_alt_stack)
+        {
+            uint64_t stacktop = (uint64_t)altstack->ss_sp + altstack->ss_size;
+            altstack->ss_flags |= SS_ONSTACK;
```

Comment:
Can you add a comment here on why adding this flag is needed during handling signal on alternate stack.

Reply:
Added

Reply:
Thanks. This is also checked in `myst_signal_altstack` if user tries to change the alternate stack while being on it.